doc/manual/compiler.rst

```diff
@@ -54,74 +76,125 @@ The Python types correspond to ARTIQ type annotations as follows:
 
-    class ProblemReturn1(EnvExperiment):
-        def build(self):
-            self.setattr_device("core")
+Kernel code can call host functions without any additional ceremony. However, such functions are assumed to return ``None``, and if a value other than ``None`` is returned, an exception is raised. To call a host function returning a value other than ``None`` its return type must be annotated, using the standard Python syntax, e.g.: ::
 
-        @kernel
-        def run(self):
-            # results in memory corruption
-            return func([1, 2, 3])
+    def return_four() -> TInt32:
+        return 4
+
+Kernels can freely modify attributes of objects shared with the host. However, by necessity, these modifications are actually applied to local copies of the objects, as the latency of immediate writeback would be unsupportable in a real-time environment. Instead, modifications are written back *when the kernel completes;* notably, this means RPCs called by a kernel itself will only have access to the unmodified host version of the object, as the kernel hasn't finished execution yet. In some cases, accessing data on the host is better handled by calling RPCs specifically to make the desired modifications. 
+
+.. warning::
+
+    Kernels *cannot and should not* return lists, arrays, or strings they have created, or any objects containing them; in the absence of a heap, the way these values are allocated means they cannot outlive the kernels they are created in. Trying to do so will normally be discovered by lifetime tracking and result in compilation errors, but in certain cases lifetime tracking will fail to detect a problem and experiments will encounter memory corruption at runtime. For example: :: 
+
+        def func(a):
+            return a
 
-This results in memory corruption at runtime.
+        class ProblemReturn1(EnvExperiment):
+            def build(self):
+                self.setattr_device("core")
+
+            @kernel
+            def run(self):
+                # results in memory corruption
+                return func([1, 2, 3])
+
+    will compile, **but corrupts at runtime.** On the other hand, lists, arrays, or strings can and should be used as inputs for RPCs, and this is the preferred method of returning data to the host. In this way the data is inherently read and sent before the kernel completes and there are no allocation issues. 
+
+Available built-in functions 
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^ 
+
+ARTIQ makes various useful built-in and mathematical functions from Python, NumPy, and (if present) SciPy available in kernel code. They are not guaranteed to be perfectly equivalent to their host equivalents (for example, ``numpy.rint`` normally rounds-to-even, but in kernel code rounds toward zero) but their behavior should basically predictable. 
+
+
+.. list-table::
```

Comment:
I don't really like this flat list, as it's disconnected from the code and potentially painful to maintain (not to mention format) but I do think it should be possible to find out which functions are available somewhere in the documentation, as opposed to "read the source code" or "just try it and see what breaks." That said, very open to suggestions on a better way to present this information.  (And is it correct to link to the NumPy/SciPy documentation? As someone not particularly used to those libraries, it wouldn't be intuitively clear to me what "j0" or "expm1" means without some kind of reference...)

Reply:
Disclaimer: as the software is in beta this list may be non-exhaustive or there may be discrepancies?

Reply:
As far as I can tell from commit histories etc. this has been pretty static since introduced four years ago, I'm not sure if that's true as such. But a disclaimer probably couldn't hurt

Reply:
Race condition: commented before I saw yours. Makes sense